crates/web5/src/credentials/verifiable_credential_1_1.rs

```diff
@@ -128,85 +118,100 @@ pub struct VerifiableCredential {
-        deserialize_with = "deserialize_option_system_time"
-    )]
-    issuance_date: Option<SystemTime>,
-    #[serde(
-        rename = "expirationDate",
-        serialize_with = "serialize_option_system_time",
-        deserialize_with = "deserialize_option_system_time"
-    )]
-    expiration_date: Option<SystemTime>,
-    #[serde(skip_serializing_if = "Option::is_none", rename = "credentialSubject")]
-    credential_subject: Option<CredentialSubject>,
+#[derive(Default)]
+pub struct VerifiableCredentialCreateOptions {
+    pub id: Option<String>,
+    pub context: Option<Vec<String>>,
+    pub r#type: Option<Vec<String>>,
+    pub issuance_date: Option<SystemTime>,
+    pub expiration_date: Option<SystemTime>,
 }
 
 impl VerifiableCredential {
-    pub fn new(
-        id: String,
-        context: Vec<String>,
-        r#type: Vec<String>,
+    pub fn create(
         issuer: Issuer,
-        issuance_date: SystemTime,
-        expiration_date: Option<SystemTime>,
         credential_subject: CredentialSubject,
-    ) -> Self {
-        let context_with_base = std::iter::once(BASE_CONTEXT.to_string())
-            .chain(context.into_iter().filter(|c| c != BASE_CONTEXT))
-            .collect::<Vec<_>>();
+        options: Option<VerifiableCredentialCreateOptions>,
+    ) -> Result<Self> {
+        if issuer.to_string().is_empty() {
+            return Err(Web5Error::Parameter(String::from(
+                "issuer id must not be empty",
+            )));
+        }
 
-        let type_with_base = std::iter::once(BASE_TYPE.to_string())
-            .chain(r#type.into_iter().filter(|t| t != BASE_TYPE))
-            .collect::<Vec<_>>();
+        if let Issuer::Object(ref named_issuer) = issuer {
+            if named_issuer.name.is_empty() {
+                return Err(Web5Error::Parameter(String::from(
+                    "named issuer name must not be empty",
+                )));
+            }
+        }
+
+        if credential_subject.to_string().is_empty() {
+            return Err(Web5Error::Parameter(String::from(
+                "subject id must not be empty",
+            )));
+        }
+
+        let options = options.unwrap_or_default();
+
+        let context = {
+            let mut contexts = options
+                .context
+                .unwrap_or_else(|| vec![BASE_CONTEXT.to_string()]);
+
+            if !contexts.contains(&BASE_CONTEXT.to_string()) {
+                contexts.insert(0, BASE_CONTEXT.to_string());
+            }
+
+            contexts
+        };
+
+        let r#type = {
+            let mut types = options
+                .r#type
+                .unwrap_or_else(|| vec![BASE_TYPE.to_string()]);
+
+            if !types.contains(&BASE_TYPE.to_string()) {
+                types.insert(0, BASE_TYPE.to_string());
+            }
 
-        Self {
-            context: context_with_base,
+            types
+        };
+
+        let id = options
```

Comment:
<img width="828" alt="image" src="https://github.com/user-attachments/assets/7efb51df-818b-4daa-aab2-3608db0325cd">
now we get into the philosophical discussion on "what it is to be a vc"

Should we let people put anything they want and create "invalid" vcs

Or should add stuff like this to a validate at the end and "dont let them create" whatever they want

Reply:
I definitely lean towards being more constrictive than not, but I think we've struck the right amount

[the web5 spec](https://tbd54566975.github.io/web5-spec/#verifiable-credentials-v11-data-model) recommends a format:

> A URI representing a unique identifier for the credential. RECOMMENDED to be of form urn:uuid:3978344f-8596-4c3a-a978-8fcaba3903c5

so we generate in accordance with that format

I'd be down to do more validation, but idk what ruleset we could add here other than "non-empty". `MUST NOT have more than one value` is impossible to validate AFAIK. `MUST be a URI` could be validated, but that seems heavy handed.

it's great food for thought! I welcome any additional ideas!